### PR TITLE
storage: truncate aggregated time series before storing

### PR DESCRIPTION
**Pull request by jd**
The current truncation code in AggregatedTimeSerie truncate on object
initialize only. That means that split up to POINTS_PER_SPLIT points are
stored, and that data is read and _truncated_ at this point.

When new points are added (existing merge), it's possible that the number of
points grows more than the number of points defined by the archive policy,
storing too many points. That's hidden because when re-read, the
AggregatedTimeSerie is truncated again.

However, that makes the storage driver store too much data.

This patches remove the max_size attribute from carbonara.AggregatedTimeSerie
and simply expose a truncate() method that can be used to truncate each split
based on the oldest point that is defined by the archive policy.

### CODE REVIEW

**Comment by jd** [comment]
Pull-request updated, HEAD is now 3f9ddedddad21f873903f3d15d8c7f914ec2d7d4

**Comment by jd** [comment]
Fix Ceph

**Comment by jd** [comment]
pep8

**Comment by jd** [comment]
Fixed truncate

**Comment by jd** [comment]
Fix resize AP case

**Comment by jd** [comment]
Fix timespan infinite case

**Comment by jd** [comment]
@gnocchixyz/devs So this patch has a great side effect for now: it implements data retention in archive policy as user expected them, i.e. "timespan=1h granularity=10min" will keep 6 points only if there are the first and last are in the same hour.

This comes by the fact that truncation is done per split, and each split does not know the global number of point, but it knows the oldest timestamp it should keep.

This overall feels safer and more inline with user expectations of AP, but it also changes the data returned.

The upgrade tests fail because of that: they used to have 10 hours of 1 second granularity (1 measure every 10 s) and now Gnocchi only returns the last hour with 1 second granularity over 10 seconds, which is more consistent with what's requested.

Now, either we close our eyes and go back to the old system and I amend this patch to cruft around this, or we honor the archive policy contract as expected. How do you feel?

**Comment by chungg** [comment]
i didn't look at code, but i know we store more than policy in ceph[#269]. is this the case for the other drivers too? don't they read the entire object and build timeseries accordingly? do you know how much excess data is stores compared to actual policy?

tbh, i don't really mind having it store more than policy as long as it adheres to policy from user pov and it's not excessive waste of space.

**Comment by jd** [comment]
> i didn't look at code, but i know we store more than policy in ceph[#269]. is this the case for the other drivers too?

Yes, because the splits are never rewritten and are only deleted every 3600 points.
Basically on a 5000 seconds 1 point per second policy, you store 3600 seconds in a first split, then start filling a second split up to 3600.
The first split is not to be deleted because you know it contains 1400 useful points. But that's ok, as you said, we truncate on __init__ so that works.

The thing is we also do that when an archive policy defines <= 3600 points: the new computed split is merged with the existing one, but there is no truncation done. This patch fixes that.

> don't they read the entire object and build timeseries accordingly? do you know how much excess data is stores compared to actual policy?

Up to 3600 (POINTS_PER_SPLIT) minus the requested data.

> tbh, i don't really mind having it store more than policy as long as it adheres to policy from user pov and it's not excessive waste of space.

Everything is relative :) If you ask for 365 points (a year) and up with 10 years, for 8 agg methods and 1 M metrics, that's 178 GB used for nothing.

So that patches fixes that but it *also* implements this by respecting `archive_policy.timespan` rather than `archive_policy.points`. I think it's the right thing to do (as if I'd do a v2 API, I'd do that) because that's what every user means when they configure an AP. But I'm requesting opinions of others before fixing the tests in that regard.

**Comment by chungg** [comment]
> Everything is relative :) If you ask for 365 points (a year) and up with 10 years, for 8 agg methods and 1 M metrics, that's 178 GB used for nothing.

yeah, that's the reason i made comment because if the overall storage (agg/unagg/raw) usage is 10PB, do i really care if it's 10PB or 10.2PB... was curious what the ratio is.

i'll look at code sometime. too lazy :P

**Comment by jd** [comment]
Fix upgrade tests

**Comment by jd** [comment]
Fix upgrade tests

File: gnocchi/tests/test_carbonara.py [hunk 1]
```diff
@@ -349,7 +349,9 @@ def test_max_size(self):
              datetime64(2014, 1, 1, 12, 0, 4),
              datetime64(2014, 1, 1, 12, 0, 9)],
             [3, 5, 6])
-        ts = self._resample(ts, numpy.timedelta64(1, 's'), 'mean', max_size=2)
+        ts = self._resample(ts, numpy.timedelta64(1, 's'), 'mean')
```
[the remark this answers is not in the document]
Comment (jd) [reply gnocchi/tests/test_carbonara.py]: Done

File: gnocchi/tests/test_carbonara.py [hunk 2]
```diff
@@ -373,7 +375,9 @@ def test_down_sampling_with_max_size(self):
              datetime64(2014, 1, 1, 12, 1, 9),
              datetime64(2014, 1, 1, 12, 2, 12)],
             [3, 5, 7, 1])
-        ts = self._resample(ts, numpy.timedelta64(60, 's'), 'mean', max_size=2)
+        ts = self._resample(ts, numpy.timedelta64(60, 's'), 'mean')
```
[the remark this answers is not in the document]
Comment (jd) [reply gnocchi/tests/test_carbonara.py]: done

File: gnocchi/tests/test_carbonara.py [hunk 3]
```diff
@@ -386,7 +390,9 @@ def test_down_sampling_with_max_size_and_method_max(self):
              datetime64(2014, 1, 1, 12, 1, 9),
              datetime64(2014, 1, 1, 12, 2, 12)],
             [3, 5, 70, 1])
-        ts = self._resample(ts, numpy.timedelta64(60, 's'), 'max', max_size=2)
+        ts = self._resample(ts, numpy.timedelta64(60, 's'), 'max')
```
[the remark this answers is not in the document]
Comment (jd) [reply gnocchi/tests/test_carbonara.py]: done

**Comment by jd** [comment]
Renamed tests and removed TODO comment

**Pull request merged**